tests/posix/common/src/pthread.c

```diff
@@ -465,6 +465,47 @@ ZTEST(pthread, test_pthread_cleanup)
 	zassert_ok(pthread_join(th, NULL));
 }
 
+static void *test_pthread_cancel_fn(void *arg)
+{
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL));
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	/* this will mark it pending */
+	pthread_cancel(pthread_self());
+
+	/* enable the thread to be cancelled */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL));
+
+	/* this should terminate the thread*/
+	pthread_testcancel();
+
+	/*
+	 * We use this large delay to indicate that the thread in fact did not exit via
+	 * pthread_testcancel(), as expected. I.e. it should not take > 200 ms to join thread
+	 */
+	zassert_ok(k_msleep(200));
+
+	return NULL;
+}
+
+ZTEST(pthread, test_pthread_testcancel)
+{
+	uint32_t now;
+	uint32_t then;
+	pthread_t th;
+
+	then = k_uptime_get_32();
+	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
+	zassert_ok(k_msleep(50));
+	zassert_ok(pthread_join(th, NULL));
+	now = k_uptime_get_32();
+
+	/* should not take > 100 ms to join thread */
+	zassert_true(now - then <= 100);
+}
```

Comment:
```suggestion
static bool testcancel_failed;

static void *test_pthread_cancel_fn(void *arg)
{
	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL));

	/* this should be ignored */
	pthread_testcancel();

	/* this will mark it pending */
	pthread_cancel(pthread_self());

	/* enable the thread to be cancelled */
	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL));

  testcancel_failed = false;

	/* this should terminate the thread*/
	pthread_testcancel();

	testcancel_failed = true;

	return NULL;
}

ZTEST(pthread, test_pthread_testcancel)
{
	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
	zassert_ok(pthread_join(th, NULL));
	zassert_false(testcancel_failed);
}
```